### DISCUSSION TITLE
Welcome to Practical-Laravel Discussions!

### DISCUSSION THREAD

**Discussion by danielgara** [PracticalBooks/Practical-Laravel#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Aot-Abderrahamane** [comment]
hello I just started your practical book, I'm grasping a lot of information by the way you explain things, but I'm still confused wondering why you didn't `use Route::resource('/','App\Http\Controllers\HomeController@index')->name("home.index");`  instead of `Route::get('/','App\Http\Controllers\HomeController@index')->name("home.index");` or I'm missing something?

**Reply by danielgara** [reply]
Hi @arthur1997-wq 

Laravel provides several ways to handle routing.

**1)** `Route::resource` -> it is used when you have a Controller with standard CRUD operations. So, if you have UserController with several methods which handles create user, edit, user, delete user and show user; then, you can use `Route::resource('users', UserController::class);` -> check that we link the entire controller class (as the second parameter), and we don't link a specific controller method. Since HomeController doesn't have CRUD operations we don't suggest to use `Route::resource` (more info here: [https://laravel.com/docs/9.x/controllers#resource-controllers](https://laravel.com/docs/9.x/controllers#resource-controllers)).

**2)** `Route::get` -> it is used when you want to link a specific route with a specific function or controller method (which is our case). In the previous case, when you access from the browser to "/" route, it will execute the HomeController index method code. (more info here: [https://laravel.com/docs/9.x/routing#the-default-route-files](https://laravel.com/docs/9.x/routing#the-default-route-files)).

In fact, Laravel provides other several ways to handle routing. You can find it here [https://laravel.com/docs/9.x/routing](https://laravel.com/docs/9.x/routing). If you want to develop a clean application, try to use only one way of those provided, or the minimum quantity. It will make easy to replicate the application code. Besides, remember to link routes only to controllers. And not to put application code inside your route files.

Let me know if that answer your question, or if you have another comment. Thanks!

**Reply by Aot-Abderrahamane** [reply]
Thank you so much you have sufficed and fulfilled

**Comment by Aot-Abderrahamane** [comment]
i have another issue about pictures i have changed the name of project example-app to OnlineStore but images won't appear for some reason I don't know `<img src="{{ asset('/storage/'.$product->getImage()) }}" `i have tried adding new products with new images but didn't show up. what do you think that cause such a problem like this?
and thanks in advance

**Reply by danielgara** [reply]
Hi @arthur1997-wq 

**1)** When you create a new product (with uploading an image), the image should appear in this folder "onlineStore/storage/app/public/*" (first, check this).
**2)** Recreate the symbolic link with `php artisan storage:link`
**3)** Finally, access the product page and you should see the images. If they don't appear, access to the browser console to check if there is an error or something. And paste it here.

Best regards.

**Reply by Aot-Abderrahamane** [reply]
Hi Mr.@danielgara 
 I have created a new product with picture, it has been added to the folder "onlineStore/storage/app/public/" then executed the commande `php artisan storage:link` they didn't appear instead it shows in console `GET http://127.0.0.1:8000/storage/3.jpg 404 (Not Found)`

**Reply by danielgara** [reply]
Hi @arthur1997-wq 

Can you check the `onlineStore/public/` folder?. It must contains a shortcut to the "storage" folder. Check the next image:

[https://i.ibb.co/cYkdc1h/Screenshot-2022-03-29-080745.jpg](https://i.ibb.co/cYkdc1h/Screenshot-2022-03-29-080745.jpg)

If it looks like a normal folder (not shortcut), please remove it and generate the storage link again. Let me know!

**Reply by danielgara** [reply]
Oh, and I forgot, try to upload .jpg images. Some other formats don't work.

**Reply by Aot-Abderrahamane** [reply]
Hi Mr. @danielgara 
its look like the same folder icon that you have in picture, and all pictures are in jpg format .
I have re-named the project back to example-app and the pictures show up but I don't understand why the project name effects them somehow

**Reply by danielgara** [reply]
@arthur1997-wq mmm, maybe it is because the storage shortcut is broken. I mean, when you renamed the project folder, the storage shortcut cannot link to the proper destination (because the destination changed).

Maybe you could try the next steps:
1) Change the project root folder name.
2) Remove the storage shortcut (remove this folder `onlineStore/public/storage`).
3) Execute the `php artisan storage:link` command again.

Maybe that way, the proper storage shortcut is generated and you can access the images.

**Reply by Aot-Abderrahamane** [reply]
It works yay, thank you so much for your time. I really appreciate it 
and your book is one of the best practical books I've ever used.

**Reply by danielgara** [reply]
@arthur1997-wq you're welcome. If you can, try to leave an honest review once you finish the book. That would be very useful for us. Thanks!

**Comment by taibikarim** [comment]
Hello. Karim from Morocco / Africa here.
What a joy to read your book. It was clear.
However, I would like to propose that you add line number to the code.
Thank you.
By the way, you have a friend in Morocco.

**Reply by danielgara** [reply]
Hello @taibikarim , thank you so much for your suggestion and your comment. I will consider it for future updates. You have also a friend in Colombia. If you can leave an honest review on Amazon that would be super useful for us. Have a great day!

**Comment by developer-luis** [comment]
Buen día ingeniero Daniel.
Segui todo el libro paso a paso de manera detalla pero a la hora de desplegar la app en Heroku tengo  un problema al intentar registrar un usuario, about y al entrar al area de pedidos, me lanza un (500 - Server Error):

Log Heroku:
....
fwd="190.0.245.42" dyno=web.1 connect=0ms service=0ms status=200 bytes=26131 protocol=http

2022-10-04T13:37:00.408368+00:00 app[web.1]: 10.1.84.207 - - [04/Oct/2022:13:37:00 +0000] "GET / HTTP/1.1" 200 3087 "http://arcane-reef-73108.herokuapp.com/register" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0

2022-10-04T13:37:00.529816+00:00 app[web.1]: 10.1.57.88 - - [04/Oct/2022:13:37:00 +0000] "GET /css/app.css HTTP/1.1" 304 - "http://arcane-reef-73108.herokuapp.com/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0

2022-10-04T13:37:00.601785+00:00 app[web.1]: 10.1.84.207 - - [04/Oct/2022:13:37:00 +0000] "GET /img/safe.png HTTP/1.1" 200 19240 "http://arcane-reef-73108.herokuapp.com/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0

2022-10-04T13:37:00.671736+00:00 app[web.1]: 10.1.90.3 - - [04/Oct/2022:13:37:00 +0000] "GET /img/game.png HTTP/1.1" 200 25896 "http://arcane-reef-73108.herokuapp.com/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0

2022-10-04T13:37:00.678396+00:00 app[web.1]: 10.1.34.133 - - [04/Oct/2022:13:37:00 +0000] "GET /img/submarine.png HTTP/1.1" 200 24330 "http://arcane-reef-73108.herokuapp.com/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0

2022-10-04T13:37:00.604230+00:00 heroku[router]: at=info method=GET path="/img/safe.png" host=arcane-reef-73108.herokuapp.com request_id=20371b9d-4937-4e01-a9ac-735630b4b4b3 fwd="190.0.245.42" dyno=web.1 connect=0ms service=0ms status=200 bytes=19475 protocol=http

2022-10-04T13:37:03.820476+00:00 app[web.1]: 10.1.90.3 - - [04/Oct/2022:13:37:03 +0000] "GET /products HTTP/1.1" 500 6609 "http://arcane-reef-73108.herokuapp.com/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0

2022-10-04T13:37:03.820605+00:00 heroku[router]: at=info method=GET path="/products" host=arcane-reef-73108.herokuapp.com request_id=842a5b08-2da2-41a7-aeea-894f017dce59 fwd="190.0.245.42" dyno=web.1 connect=0ms service=8ms status=500 bytes=7733 protocol=http

2022-10-04T13:37:06.272895+00:00 heroku[router]: at=info method=GET path="/favicon.ico" host=arcane-reef-73108.herokuapp.com request_id=d73e6c08-472f-4df1-bdfe-759809a64c5f fwd="190.0.245.42" dyno=web.1 connect=0ms service=0ms status=200 bytes=231 protocol=http

2022-10-04T13:37:06.273101+00:00 app[web.1]: 10.1.90.3 - - [04/Oct/2022:13:37:06 +0000] "GET /favicon.ico HTTP/1.1" 200 - "http://arcane-reef-73108.herokuapp.com/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0

**Reply by developer-luis** [reply]
SOLUCION: quitar el .env del gitignore

**Reply by danielgara** [reply]
@developer-luis que bueno que lo solucionaste! Si de pronto terminaste el libro y puedes dejar una reseña honesta en Amazon, eso nos ayudaría un montón. Un feliz día.

**Comment by meticulouschris** [comment]
In chapter 11, toward the end it tells us to copy/paste the SQL query in figure 11-3. I typed it in manually because I couldn't find that code in the github repo.

**Reply by danielgara** [reply]
Hi @meticulouschris thanks for the report. You were right! I forgot to add that query text.

I just included it here: https://github.com/PracticalBooks/Practical-Laravel/blob/main/Chapter11/onlineStore/CHAPTER-11-SQL-QUERY.txt

Thanks again!

**Comment by CarloCanenero** [comment]
Hello, i'm at the end of the book but Heroku now is a paid service.
Can you provide an alternative and the steps to deploy the project?
Unfortunately the steps in the book are not valid anymore, thank you very much! :)

**Reply by danielgara** [reply]
Hi @CarloCanenero - I will make a research (it is a shame that Heroku doesn't offer a free service anymore), I found a couple of free tools but they require to use Docker, so the tutorial must be completely changed. 

Once I have some news I will publish it here.

**Reply by CarloCanenero** [reply]
Thank you very much! :)
